website/blog/how-to-troubleshoot-w-bisect.md

```diff
@@ -0,0 +1,407 @@
+```
+
+**Explanation**: The last-known good release for us was `8.0.1`, and the bad release was `8.1.0`. Bazelisk will grab the list of commits between 2 versions from the Github API and start a bisect. For each bisect commit, it will grab a pre-built Bazel binary of that version from a Google Cloud Storage bucket and run the test command accordingly. The `.bazelversion` file will be ignored in these runs.
+
+Because the issue was related to external dependencies download, we want to have a clean workspace for each test run to avoid local cache affecting the test result. Instead of running `bazel clean --expunge` manually, we leverage `--output_base=$(mktemp -d)` here to create a temporary directory for each bisect run instead.
+
+Since we know that the issue was caused by the external download of `@rules_cc`, we can narrow down the test command to a `bazel query` that would still trigger the same download. This should help speed up each bisect run and reduce the chance of flakiness.
+
+```bash
+$ bazelisk --bisect=8.0.1..8.1.0 --output_base=$(mktemp -d) query --config=remote-minimal @rules_cc//:all
+```
+
+Note that using a separate output base directory will spin up a new Bazel JVM process each time. So if the bisect range is large, you may run into memory issues.
+
+And here is the result:
+
+```bash
+--- Getting the list of commits between 8.0.1 and 8.1.0
+Found 95 commits between (24cba660a7231786405ac40335c2e7e5bd4d6859, 8.1.0]
+--- Verifying if the given good Bazel commit (24cba660a7231786405ac40335c2e7e5bd4d6859) is actually good
+--- Start bisecting
+--- Testing with Bazel built at c5cf63199b3964b4a188da73a6c24599468131e9, 95 commits remaining...
+--- Succeeded at c5cf63199b3964b4a188da73a6c24599468131e9
+--- Testing with Bazel built at 040769767613287a0f1b5ceed41b9d7729126983, 47 commits remaining...
+
+2025/02/18 10:56:29 Using unreleased version at commit 040769767613287a0f1b5ceed41b9d7729126983
+2025/02/18 10:56:29 Downloading https://storage.googleapis.com/bazel-builds/artifacts/macos_arm64/040769767613287a0f1b5ceed41b9d7729126983/bazel...
+2025/02/18 10:56:29 Skipping basic authentication for storage.googleapis.com because no credentials found in /Users/sluongng/.netrc
+2025/02/18 10:56:29 could not run Bazel: could not download Bazel: failed to download bazel: failed to download bazel: HTTP GET https://storage.googleapis.com/bazel-builds/artifacts/macos_arm64/040769767613287a0f1b5ceed41b9d7729126983/bazel failed with error 404
+```
+
+Here, Bazelisk was able to identify `8.0.1` to be `24cba660a` and verified that it was indeed a good commit. It then started bisecting the commits between `24cba660a` and `8.1.0`. The first bisect commit `c5cf63199b` was good. However, the second bisect commit `0407697676` failed to download the Bazel binary from the Google Cloud Storage (GCS) bucket. This is a different failure than the original issue we are investigating, so how do we proceed?
```

Comment:
Could you report this as a bug to Bazelisk? It should skip these versions instead of failing the bisect.

Reply:
Sent https://github.com/bazelbuild/bazelisk/issues/663